analyses/npl/01_get_stations.md

```diff
@@ -0,0 +1,379 @@
+with pd.ExcelWriter(STATION_OUTPUT) as writer:
+    for basin in basin_list:
+        df = df_dhm[df_dhm.geometry.within(basin_shape_dict[basin])]
+        df.to_excel(writer, sheet_name=basin, index=False)
+```
+
+## Get final list of stations to put into config file
+
+```python
+# Based on email exchange with Ragindra
+# Using GloFAS ID since two stations in Koshi have exactly the same name
+stations_final = {
+    'koshi': [4475, 4619, 4425, 4403],
+    'karnali': [4385, 915, 4469, 4393],
+    'rapti': [4456],
+    'bagmati': [4399],
+    'babai': [4416]
+}
+stations_final_yaml = {}
+for basin in stations_final.values():
+    for glofas_id in basin:
+        row = df_dhm.loc[df_dhm['id_glofas'] == f'G{str(glofas_id).zfill(4)}'].iloc[0]
+        stations_final_yaml[row['name_glofas']] = {
+            'lat': row['lat_glofas'],
+            'lon': row['lon_glofas']
+        }
+
+        
+with open(STATION_OUTPUT_YAML, 'w') as f:
+    yaml.dump(stations_final_yaml, f)
+```
+
+### Create final points shapefile for making map visuals
+
+```python
+sheet_id = "1I3vszdCDDxFnlhieywY2C9MbU-0TmGVn"
+url = "https://docs.google.com/spreadsheets/d/{0}/gviz/tq?tqx=out:csv&sheet={1}"
+
+
+  
+for source in ['glofas', 'dhm']:
+    df_final = gpd.GeoDataFrame()
+    for sheet_name in ['Koshi', 'Karnali', 'West_Rapti', 'Bagmati', 'Babai']:
+        df = pd.read_csv(url.format(sheet_id, sheet_name)).dropna(subset=['name_glofas'])
+        df = gpd.GeoDataFrame(df[[f'name_{source}', f'lat_{source}', f'lon_{source}']])
+        df_final = df_final.append(df, ignore_index=True)
+    df_final['geometry'] = gpd.points_from_xy(df_final[f'lon_{source}'], df_final[f'lat_{source}'])
+    df_final = df_final.set_crs("EPSG:4326")
+    df_final.to_file(STATIONS_FINAL[source], driver="GPKG", index=False)
+
+```
+
+# Appendix
+
+
+### Try to connect gov with dhm 
+
+```python
+# Match it to the gov hydrology stations
+# It would be more proper to do a merge but it's hard with two possible columns
+
+df_dhm = df_dhm.dropna(subset=['river_dhm'])
+df_dhm['gov_name'] = ''
+for i, row in df_dhm.iterrows():
+    gov_name_1 = row['river_dhm'].split('River')[0].strip() + ' at ' + row['name_dhm']
+    gov_name_2 = row['river_dhm'] + ' at ' + row['name_dhm']
+    for gov_name in gov_name_1, gov_name_2:
+        matches = df_gov['name'] == gov_name
+        if sum(matches):
+            df_dhm.at[i, 'gov_name'] = gov_name
+```
+
+```python
+match_dict = {
+    'Jamu': 'Jamu, Dailekh',
+    'Kusum': 'Rapti River at Kusum (Velocity)',
+    'Borlangpul': 'Aadhi Khola at Borlangpul',
+    'Angsing': 'Jyagdi Khola at Dangsing',
+    'Bimalnagar': 'Marsyangdi at Bimalnagar',
+    'Betrawati': 'Trishuli at Betrawati',
+    'Rajaiya': 'Rajaiya (rainfall station)',
+    'Lothar': 'Lothar Khola at Lothar',
+    'Jalbire': 'Balefi at Jalbire',
+    'Rasnalu': 'KhimtiKhola at Rasnalu',
+    'Chatara': 'Saptakoshi at Chatara (old)'
+}
+```
+
+I give up, will find another way to get the basin
```

Comment:
Hahaha...

Reply:
Not sure if it's helpful at this point but we used this for fuzzy matching of admin area names in Bangladesh: https://github.com/seatgeek/fuzzywuzzy

Reply:
Ah great, for next time ;)